### ISSUE TITLE
Unpinning fails until the app is relaunched

### ISSUE THREAD

**Issue by sobri909** [ParsePlatform/Parse-SDK-iOS-OSX#768]
Although the unpin callback reports success, the object is still returned in queries for that pin. It's not until the app is killed and relaunched that attempts to unpin that object will truly succeed.

Here's the method I'm finding this problem in: 

```swift
    func flushActivityEvents() {
        log()

        if flushing {
            log("already trying to flush")
            return
        }

        if beingFlushed.count >= beingFlushedLimit {
            log("already at flush limit")
            return
        }

        flushing = true

        // get all unsaved ActivityEvents
        let query = ActivityEvent.query()!
        query.fromPinWithName(ActivityEvent.UnsavedPin)

        query.findObjectsInBackgroundWithBlock { (results: [PFObject]?, error: NSError?) in
            if error != nil {
                log("error: \(error!)")
            }

            log("results: \(results!.count)")

            for event in results as! [ActivityEvent] {
                if self.beingFlushed.count >= self.beingFlushedLimit {
                    log("reached flush limit")
                    break
                }

                // don't double flush
                if self.beingFlushed.indexOf(event) != nil {
                    continue
                }

                if !event.dirty {
                    log("not dirty. unpinning and skipping")
                    event.unpinInBackgroundWithName(ActivityEvent.UnsavedPin)
                    continue
                }

                self.beingFlushed.append(event)

                // save it, then remove it from the "unsaved" pin on success
                event.saveInBackgroundWithBlock() { (succeeded: Bool, error: NSError?) in
                    self.beingFlushed.remove(event)
                    if succeeded {
                        event.unpinInBackgroundWithName(ActivityEvent.UnsavedPin)
                        log("unpinned: \(event.objectId) from pin: \(ActivityEvent.UnsavedPin)")
                    }
                }
            }

            self.flushing = false
        }
    }
```

`ActivityEvent.UnsavedPin` is defined thus: 

```swift
static let UnsavedPin = "UnsavedPin"
```

The objects are also stored under a second pin, at creation time: 

```swift
event.pinInBackgroundWithName(ActivityEvent.CachePin)
event.pinInBackgroundWithName(ActivityEvent.UnsavedPin)
```

I'm using an explicit `CachePin` just as a way to try to debug the problem. I thought that maybe the default pin was acting differently from others. But the explicit pin name didn't change anything.

So what's happening is my app creates a new `ActivityEvent` every 10 seconds, then (if plugged into power) calls `flushActivityEvents()`. On each call to this method, the result count is +1, when it should be exactly 1 every time (provided the saves are succeeding, which they are). So the number of results increases by 1 every ten seconds, instead of staying a steady 1. 

The two logging lines that show the problem are: 

```swift
log("unpinned: \(event.objectId) from pin: \(ActivityEvent.UnsavedPin)")
```

and

```swift
log("not dirty. unpinning and skipping")
```

The second one shouldn't be possible, because if the object isn't dirty (the previous save succeeded), then it shouldn't be in the pin, and shouldn't be returned in the results. But it is. 

Also that second attempt to unpin it also achieves nothing. 

If the app is killed and relaunched, then the first call to `flushActivityEvents()` will unpin all the ones that aren't dirty, and **that** will succeed. But from then each new object created will fail to be unpinned. 

Thoughts? Am I missing something obvious? Have I misunderstood pinning?

**Comment by sobri909** [comment]
As an aside, this code snippet in the iOS guide appears to be wrong: 

https://parse.com/docs/ios/guide#local-datastore-syncing-local-changes

It queries an explicit pin thus: 

```swift
query.fromPinWithName("MyChanges")
```

But then unpins from the default pin, thus: 

```swift
return score.unpinInBackground()
```

That, to my eyes, looks like it shouldn't work. It would still remain in the `MyChanges` pin, thus would continue to be returned in results. Unless I really am completely misunderstanding pins.

**Comment by sobri909** [comment]
I'm also seeing the problem in another chunk of code, for another class. When my Processor decides that some objects should be deleted, it doesn't do the delete until the device is plugged back into power. So it does this: 

```swift
        // betweener and newest are deadmen
        betweener.pinInBackgroundWithName(HistoryItem.DeadmanPin)
        betweener.unpinInBackgroundWithName(HistoryItem.UnsavedPin)
        newest.pinInBackgroundWithName(HistoryItem.DeadmanPin)
        newest.unpinInBackgroundWithName(HistoryItem.UnsavedPin)

        // betweener and newest shouldn't come back in search results
        betweener.unpinInBackgroundWithName(HistoryItem.CachePin)
        newest.unpinInBackgroundWithName(HistoryItem.CachePin)
```

But the query that pulls these objects for display in the UI, which has this clause, continues to get these objects returned in the results: 

```swift
        if !fromRemote {
            itemsQuery.fromPinWithName(HistoryItem.CachePin)
        }
```

Those objects don't stop returning in the results until they really are deleted (once plugged into power). So it's the same problem of the unpinning not actually succeeding, even though it reports success.

**Comment by nlutsenko** [comment]
Hey @sobri909, few things here that might help clarify the behavior:
- With Local Datastore - there is a guarantee that there is going to be 1 and only 1 instance of any PFObject for a given className/objectId pair in memory
- Pinning the same object to multiple pins will simply update the reference
- Unless there are no references to an object - it's not fully evicted from LDS and is queryable from pins that still have that object referenced

The code above makes sense and looks like you are familiar with this already, but unfortunately I wasn't able to reproduce the issue locally.
Could you attach a piece of sample code/project that clearly reproduces the issue? It will help a lot in faster turnaround on this issue.

P.S. You are correct in terms of documentation - I think that code snippet is incorrect. Will gladly review a pull request on Docs repo.

**Comment by sobri909** [comment]
@nlutsenko So it sounds like my understanding of how pinning works is correct. Which is good. It's meant to work the way I'd hoped it would.

Just need to find out how to reproduce the problem for you then. I'll have a stab at #772 first, as that one might be easier to make a sample project for. And is possibly related.

**Comment by nlutsenko** [comment]
Yup, sounds related. Thank you for commitment to reproduce this. It helps to make Parse better as well as helps everyone in the community.

**Comment by sobri909** [comment]
### Sample Project

[ParseTest.zip](https://github.com/ParsePlatform/Parse-SDK-iOS-OSX/files/101729/ParseTest.zip)

### Summary

This is reproducing the problem 100%, so it's proving a lot easier to debug.

The problem is caused by adding `ActivityEvent` instances to the `HistoryItem`'s `activities` array. Once that is done, even though the activity has been successfully unpinned, it will still be returned in queries for that pin. 

Setting `shouldAddActivitiesToHistoryItems` to `false` stops the problem from reproducing.

My immediate suspicion was that this was because I was erroneously using `activities` as mutable. But replacing the `append()` with this: 

```swift
        let activities = User.me.currentHistoryItem!.activities
        User.me.currentHistoryItem!.activities = activities + [activityEvent]
```

doesn't fix it. So it doesn't appear to be a mutable containers problem. 

My current suspicion is that the problem is caused by the `HistoryItem` being pinned with the same pin name as the activity events. I had assumed that pins were per object type, but from futzing around in the SQLite file today it looks like pins are singular, and contain all object types using that pin name. So my next thing to try is to use separate pin names per object type. 

### Notes

- The use of `delay()` for adding the activity to the history item on line 84 isn't relevant to much. If I do it in the same run cycle, either before or after the save call, it makes no difference. I simply did it with a delay to keep it from clashing with the saving / unpinning. But yeah, it makes no difference.

**Comment by sobri909** [comment]
Okay, using distinct pin names per object type has fixed it. (Though I'm partially of the mind that this is still a problem. I'll add some thoughts on that below).

I changed HistoryItem to this: 

```swift
class HistoryItem: PFObject, PFSubclassing {

    static let CachePin = "HistoryItemCachePin"
    static let UnsavedPin = "HistoryItemUnsavedPin"

    ...
}
```

And ActivityEvent to this: 

```swift
class ActivityEvent: PFObject, PFSubclassing {

    static let CachePin = "ActivityEventCachePin"
    static let UnsavedPin = "ActivityEventUnsavedPin"

    ...
}
```

And the problem disappears. 

So this means that if an object has a pointer to an object of another type, and that child object is unpinned, it will *remain* in that pin so long as its parent is in that pin. So a "successful" unpin is ... misleading.

This is probably by design. I think I've read somewhere that pinning is supposed to recurse the pinning down the tree to all children. So even though I've been unpinning, because it's a child of something else in the same pin, the unpinning either does nothing or the pin is reapplied afterwards by something else. 

So yeah, we're partly looking at an error on my part in misunderstanding pins. I'd a) assumed they were per object type, and b) not paid attention to the recursive nature of pinning.

But given that I fell into this trap, and it took quite a struggle to find my way out, I think there's also either a problem with either documentation or design. I haven't thought it through properly, but one way or another I shouldn't have been able to end up in this mess so easily. It needs thought.

**Comment by sobri909** [comment]
Actually, this pinning recursion, does the same hold true for unpinning?

If I pin a parent that has children, then all children will be pinned (if I'm now understanding correctly). If I then unpin the parent, will all children also be unpinned?

This could be worked around by pinning the children under a different name, for those that I want to keep around. But this conflation of manually pinned objects and automatically pinned objects (via recursive pinning) is starting to feel unintuitive. If I had manually pinned some of those children, I wouldn't expect them to become unpinned when I unpin the parent. I would expect my manual pinning to persist. 

Obviously, yes, using different pin names per object type is what I'm going to do now to work around this. But again, this feels like it needs more thought. It's slightly "thar be dragons".

**Comment by sobri909** [comment]
Ugh. By changing over to relying on automatic pinning of children I've ended up with a corrupt LDS and fetches are now producing unrecoverable errors. 

> Code=120 "This object is not available in the offline cache."

There appears to be no way to recover from this. Nor any indication of which object is at fault. 

For reference, I had changed the pinning to only re-pin the base object when any new children were added, instead of pinning the children directly. I'd assumed this would cause the pinning to cascade to all children, ensuring that all were safely pinned. It appears to be not so.
